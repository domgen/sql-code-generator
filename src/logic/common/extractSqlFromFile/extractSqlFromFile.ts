import { extractSqlFromSqlFile } from './extractSqlFromSqlFile';
import { extractSqlFromTsFile } from './extractSqlFromTsFile';

const SUPPORTED_FILE_EXTENSIONS = ['.sql', '.ts'];
export const extractSqlFromFile = async ({
  filePath,
}: {
  filePath: string;
}) => {
  // 1. grab extension from file; check that we support it
  const extension = `.${filePath.split('.').slice(-1)[0]}`;
  if (!SUPPORTED_FILE_EXTENSIONS.includes(extension)) {
    throw new Error(`file extension '.${extension}' is not supported`);
  }

  // 2. grab sql from file, based on file type
  if (extension === '.sql') return extractSqlFromSqlFile({ filePath });
  if (extension === '.ts') return extractSqlFromTsFile({ filePath });
  throw new Error('unexpected'); // should never reach here, since we checked earlier that the extension was supported
};
