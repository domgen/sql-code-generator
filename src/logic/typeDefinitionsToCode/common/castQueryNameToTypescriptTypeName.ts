import { pascalCase } from 'pascal-case';

export const castQueryNameToTypescriptTypeName = ({
  name,
}: {
  name: string;
}) => {
  const queryNameInPascalCase = pascalCase(name);
  return `SqlQuery${queryNameInPascalCase}`;
};
