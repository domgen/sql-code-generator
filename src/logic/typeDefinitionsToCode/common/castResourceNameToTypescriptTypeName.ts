import { pascalCase } from 'pascal-case';

import { ResourceType } from '../../../model';

export const castResourceNameToTypescriptTypeName = ({
  name,
  resourceType,
}: {
  name: string;
  resourceType: ResourceType;
}) => {
  const resourceTypeInTitleCase = resourceType[0].toUpperCase() + resourceType.substr(1).toLowerCase();
  const resourceNameInPascalCase = pascalCase(name);
  return `Sql${resourceTypeInTitleCase}${resourceNameInPascalCase}`;
};
