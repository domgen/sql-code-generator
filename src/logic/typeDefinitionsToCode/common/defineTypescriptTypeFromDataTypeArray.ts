import { DataType } from '../../../model';

export const defineTypescriptTypeFromDataTypeArray = ({ type }: { type: DataType[] }) => type.join(' | ');
