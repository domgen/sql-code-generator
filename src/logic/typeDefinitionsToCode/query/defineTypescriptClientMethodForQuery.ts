import { camelCase } from 'change-case';
import { castQueryNameToTypescriptTypeName } from '../common/castQueryNameToTypescriptTypeName';

/**
 * given:
 * - typescript type definition of input and output
 *
 * produce:
 * - typescript function that gets the inputs and returns promise of output
 * - typescript function that calls database in best practice way
 *   - uses yesql to pass inputs into parameterized statement
 *   - parses result and extracts output
 *   - logs inputs and output stats
 */
export const defineTypescriptClientMethodForQuery = ({
  name,
  relativePathToQueryExport,
}: {
  name: string;
  relativePathToQueryExport: string;
}) => {
  const typescriptTypeName = castQueryNameToTypescriptTypeName({ name });
  const inputTypeName = `${typescriptTypeName}Input`;
  const outputTypeName = `${typescriptTypeName}Output`;
  const typescriptClientMethodName = camelCase(typescriptTypeName); // i.e., same as typename, except camel case
  const clientMethodDefinition = `
import { mysql as prepare } from 'yesql';
import { query } from '${relativePathToQueryExport}';
import { ${inputTypeName}, ${outputTypeName} } from '../types';
import { DatabaseExecuteCommand, LogMethod } from './common';

export const ${typescriptClientMethodName} = async ({
  dbExecute,
  logDebug,
  input,
}: {
  dbExecute: DatabaseExecuteCommand;
  logDebug: LogMethod;
  input: ${inputTypeName};
}): Promise<${outputTypeName}[]> => {
  // 1. define the query with yesql
  const { sql: preparedSql, values: preparedValues } = prepare(query)(input);

  // 2. log that we're running the request
  logDebug('${typescriptClientMethodName}.input', { input });

  // 3. execute the query
  const output = await dbExecute({ sql: preparedSql, values: preparedValues });

  // 4. log that we've executed the request
  logDebug('${typescriptClientMethodName}.output', { output });

  // 5. return the output
  return output;
};
  `.trim();
  return clientMethodDefinition;
};
