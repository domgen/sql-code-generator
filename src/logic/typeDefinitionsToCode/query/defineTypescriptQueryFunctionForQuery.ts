import { camelCase } from 'change-case';
import { castQueryNameToTypescriptTypeName } from '../common/castQueryNameToTypescriptTypeName';

export interface QueryFunctionDefinition {
  code: string;
  imports: {
    queryNameAlias: string;
    generatedTypes: string[];
  };
}

/**
 * given:
 * - typescript type definition of input and output
 *
 * produce:
 * - typescript function that gets the inputs and returns promise of output
 * - typescript function that calls database in best practice way
 *   - uses yesql to pass inputs into parameterized statement
 *   - parses result and extracts output
 *   - logs inputs and output stats
 */
export const defineTypescriptQueryFunctionForQuery = ({ name }: { name: string }): QueryFunctionDefinition => {
  const typescriptTypeName = castQueryNameToTypescriptTypeName({ name });

  // define the query name alias, import { query as ... } under
  const queryNameAlias = `${camelCase(typescriptTypeName)}Sql`; // i.e., sqlQueryUpsertUserSql

  // define the generated types to import
  const outputTypeName = `${typescriptTypeName}Output`;
  const inputTypeName = `${typescriptTypeName}Input`;
  const generatedTypesToImport = [inputTypeName, outputTypeName];

  // define the function body
  const typescriptQueryFunctionName = camelCase(typescriptTypeName); // i.e., same as typename, except camel case
  const queryFunctionDefinition = `
export const ${typescriptQueryFunctionName} = async ({
  dbExecute,
  logDebug,
  input,
}: {
  dbExecute: DatabaseExecuteCommand;
  logDebug: LogMethod;
  input: ${inputTypeName};
}): Promise<${outputTypeName}[]> => {
  // 1. define the query with yesql
  const { sql: preparedSql, values: preparedValues } = prepare(${queryNameAlias})(input || {});

  // 2. log that we're running the request
  logDebug('${typescriptQueryFunctionName}.input', { input });

  // 3. execute the query
  const [output] = await dbExecute({ sql: preparedSql, values: preparedValues });

  // 4. log that we've executed the request
  logDebug('${typescriptQueryFunctionName}.output', { output });

  // 5. return the output
  return output;
};
  `.trim();
  return {
    imports: {
      queryNameAlias,
      generatedTypes: generatedTypesToImport,
    },
    code: queryFunctionDefinition,
  };
};
