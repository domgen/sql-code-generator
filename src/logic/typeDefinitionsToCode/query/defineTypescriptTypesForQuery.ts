import { TypeDefinitionOfQuery } from '../../../model/valueObjects/TypeDefinitionOfQuery';
import { castQueryNameToTypescriptTypeName } from '../common/castQueryNameToTypescriptTypeName';
import { defineTypescriptTypeFromReference } from '../common/defineTypescriptTypeFromReference/defineTypescriptTypeFromReference';
import { TypeDefinition } from '../../../model';

export const defineTypescriptTypesForQuery = ({
  definition,
  allDefinitions,
}: {
  definition: TypeDefinitionOfQuery;
  allDefinitions: TypeDefinition[];
}) => {
  // define the typescript type name for the query
  const typescriptTypeName = castQueryNameToTypescriptTypeName({ name: definition.name });

  // define the input interface
  const typescriptInputInterfacePropertyDefinitions = definition.inputVariables.map((inputVariable) => {
    const typescriptTypeForReference = defineTypescriptTypeFromReference({
      reference: inputVariable.typeReference,
      queryTableReferences: definition.tableReferences,
      typeDefinitions: allDefinitions,
    });
    return `${inputVariable.name}: ${typescriptTypeForReference};`;
  });
  const typescriptInputInterfaceDefinition = `
export interface ${typescriptTypeName}Input {
  ${typescriptInputInterfacePropertyDefinitions.join('\n  ')}
}
  `.trim();

  // define the output interface
  const typescriptOutputInterfacePropertyDefinitions = definition.selectExpressions.map((selectExpression) => {
    const typescriptTypeForReference = defineTypescriptTypeFromReference({
      reference: selectExpression.typeReference,
      queryTableReferences: definition.tableReferences,
      typeDefinitions: allDefinitions,
    });
    return `${selectExpression.alias}: ${typescriptTypeForReference};`;
  });
  const typescriptOutputInterfaceDefinition = `
export interface ${typescriptTypeName}Output {
  ${typescriptOutputInterfacePropertyDefinitions.join('\n  ')}
}
  `.trim();

  // return typescript types
  return `
${typescriptInputInterfaceDefinition}
${typescriptOutputInterfaceDefinition}
  `.trim();
};
