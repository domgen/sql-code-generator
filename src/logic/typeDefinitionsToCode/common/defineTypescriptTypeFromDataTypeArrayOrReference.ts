import { DataType, TypeDefinitionReference, TypeDefinition, TypeDefinitionOfQueryTableReference } from '../../../domain';
import { defineTypescriptTypeFromReference } from './defineTypescriptTypeFromReference/defineTypescriptTypeFromReference';
import { defineTypescriptTypeFromDataTypeArray } from './defineTypescriptTypeFromDataTypeArray';

export const defineTypescriptTypeFromDataTypeArrayOrReference = ({
  type,
  queryTableReferences,
  typeDefinitions,
}: {
  type: DataType[] | TypeDefinitionReference;
  typeDefinitions: TypeDefinition[];
  queryTableReferences: TypeDefinitionOfQueryTableReference[];
}) => {
  // if its a type reference, then use that handler
  if (type instanceof TypeDefinitionReference) {
    return defineTypescriptTypeFromReference({
      reference: type,
      queryTableReferences,
      typeDefinitions,
    });
  }

  // else, it must be data array. use that handler
  return defineTypescriptTypeFromDataTypeArray({ type });
};
