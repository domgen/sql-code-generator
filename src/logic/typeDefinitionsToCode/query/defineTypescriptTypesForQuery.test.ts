import { extractSqlFromFile } from '../../common/extractSqlFromFile';
import { extractTypeDefinitionFromQuerySql } from '../../sqlToTypeDefinitions/query/extractTypeDefinitionFromQuerySql';
import { defineTypescriptTypesForQuery } from './defineTypescriptTypesForQuery';
import { TypeDefinitionOfResourceTable, TypeDefinitionOfResourceView } from '../../../model';

describe('defineTypescriptTypesForQuery', () => {
  it('should be able to determine types accurately an example of selecting columns a single table by id', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../__test_assets__/queries/find_image_by_id.sql`,
    });
    const def = extractTypeDefinitionFromQuerySql({
      name: 'find_image_by_id',
      path: '__PATH__',
      sql,
    });
    const code = defineTypescriptTypesForQuery({
      definition: def,
      allDefinitions: [new TypeDefinitionOfResourceTable({ name: 'image', columns: [] })],
    });
    expect(code).toMatchSnapshot();
  });
  it('should be able to determine types accurately when selecting columns from multiple tables, no input variables', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../__test_assets__/queries/select_suggestion.sql`,
    });
    const def = extractTypeDefinitionFromQuerySql({
      name: 'select_suggestion',
      path: '__PATH__',
      sql,
    });
    const code = defineTypescriptTypesForQuery({
      definition: def,
      allDefinitions: [
        new TypeDefinitionOfResourceTable({ name: 'suggestion', columns: [] }),
        new TypeDefinitionOfResourceTable({ name: 'suggestion_version', columns: [] }),
      ],
    });
    expect(code).toMatchSnapshot();
  });
  it('should be able to determine types accurately an upsert query', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../__test_assets__/queries/upsert_suggestion.sql`,
    });
    const def = extractTypeDefinitionFromQuerySql({
      name: 'upsert_suggestion',
      path: '__PATH__',
      sql,
    });
    const code = defineTypescriptTypesForQuery({
      definition: def,
      allDefinitions: [],
    });
    expect(code).toMatchSnapshot();
  });
  it('should be able to determine types accurately for this example that selects from tables, functions, and views', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../__test_assets__/queries/find_users_by_last_name.sql`,
    });
    const def = extractTypeDefinitionFromQuerySql({
      name: 'find_users_by_last_name',
      path: '__PATH__',
      sql,
    });
    const code = defineTypescriptTypesForQuery({
      definition: def,
      allDefinitions: [
        new TypeDefinitionOfResourceTable({ name: 'user', columns: [] }),
        new TypeDefinitionOfResourceTable({ name: 'phone', columns: [] }),
        new TypeDefinitionOfResourceView({
          name: 'view_user_profile_current',
          selectExpressions: [],
          tableReferences: [],
        }),
      ],
    });
    expect(code).toMatchSnapshot();
  });
});
