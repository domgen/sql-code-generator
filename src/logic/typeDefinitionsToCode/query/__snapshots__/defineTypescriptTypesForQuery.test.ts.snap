// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineTypescriptTypesForQuery should be able to determine types accurately an example of selecting columns a single table by id 1`] = `
"export interface SqlQueryFindImageByIdInput {
  id: SqlTableImage['id'];
}
export interface SqlQueryFindImageByIdOutput {
  id: SqlTableImage['id'];
  url: SqlTableImage['url'];
  caption: SqlTableImage['caption'];
  alt_text: SqlTableImage['alt_text'];
  credit: SqlTableImage['credit'];
}"
`;

exports[`defineTypescriptTypesForQuery should be able to determine types accurately an upsert query 1`] = `
"export interface SqlQueryUpsertSuggestionInput {
  suggestionSource: SqlFunctionUpsertSuggestionInput['0'];
  externalId: SqlFunctionUpsertSuggestionInput['1'];
  suggestedIdeaId: SqlFunctionUpsertSuggestionInput['2'];
  status: SqlFunctionUpsertSuggestionInput['3'];
  result: SqlFunctionUpsertSuggestionInput['4'];
}
export interface SqlQueryUpsertSuggestionOutput {
  id: SqlFunctionUpsertSuggestionOutput;
}"
`;

exports[`defineTypescriptTypesForQuery should be able to determine types accurately for this example that selects from tables, functions, and views 1`] = `
"export interface SqlQueryFindUsersByLastNameInput {
  lastName: SqlTableUser['last_name'];
}
export interface SqlQueryFindUsersByLastNameOutput {
  id: SqlTableUser['id'];
  full_name: SqlFunctionConcatOutput;
  age: SqlTableUser['age'];
  phone_number: SqlTablePhone['number'];
  description: SqlViewViewUserProfileCurrent['description'];
}"
`;

exports[`defineTypescriptTypesForQuery should be able to determine types accurately when selecting columns from multiple tables, no input variables 1`] = `
"export interface SqlQuerySelectSuggestionInput {
  
}
export interface SqlQuerySelectSuggestionOutput {
  id: SqlTableSuggestion['id'];
  uuid: SqlTableSuggestion['uuid'];
  suggestion_source: SqlTableSuggestion['suggestion_source'];
  external_id: SqlTableSuggestion['external_id'];
  suggested_idea_id: SqlTableSuggestion['suggested_idea_id'];
  resultant_curated_idea_id: SqlTableSuggestion['resultant_curated_idea_id'];
  status: SqlTableSuggestionVersion['status'];
  result: SqlTableSuggestionVersion['result'];
  created_at: SqlTableSuggestion['created_at'];
  effective_at: SqlTableSuggestionVersion['effective_at'];
  updated_at: SqlTableSuggestionVersion['created_at'];
}"
`;
